hack/tools/metal.py

```diff
@@ -0,0 +1,171 @@
+   def create_device(self, project_id, name, key_id, userdata):
+      device = self.packet_manager.create_device(project_id=project_id,
+                                     hostname=name,
+                                     plan='c1.small.x86', metro='sv',
+                                     operating_system='ubuntu_18_04',
+                                     facility="ewr1", billing_cycle='hourly',
+                                     project_ssh_keys=[key_id],
+                                     userdata=userdata)
+      self.logger.info(f"created device {device.hostname}")
+
+      return device
+
+   # TODO timeouts
+   def wait_until_device_ready(self, device):
+      while True:
+         d = self.packet_manager.get_device(device_id=device.id)
+         self.logger.info(f"checking state of device {d.hostname} ...")
+         if d.state == "active":
+             self.logger.info(f"{d.hostname} running")
+             break
+         self.logger.info(f"state '{d.state}' != 'active', sleeping for 10s")
+         time.sleep(10)
+
+      ips = device.ips()
+      self.ip = ips[0].address
+
+      while True:
+         shell = self.new_shell(self.ip)
+         self.logger.info("checking userdata has completed...")
+         with shell:
+            result = shell.run(["ls", "/flintlock_ready"], allow_error=True)
+         if result.return_code == 0:
+            self.logger.info("userdata ran successfully")
+            break
+         self.logger.info("userdata still running, sleeping for 10s")
+         time.sleep(10)
+
+      self.logger.info(f"device {device.hostname} (id: {device.id}) ready")
+
+   def run_ssh_command(self, cmd, cwd):
+      shell = self.new_shell(self.ip)
+      with shell:
+         result = shell.run(command=cmd, cwd=cwd, stdout=sys.stdout.buffer, stderr=sys.stderr.buffer, allow_error=True)
+      self.logger.info("command exited with code %d", result.return_code)
+
+   def delete_all(self, project, device, key):
+      if device != None:
+         device.delete()
+         self.logger.info(f"deleted device {device.hostname}")
+
+      if key != None:
+         key.delete()
+         os.remove(self.private_key_path)
+         os.remove(self.public_key_path)
+         self.logger.info(f"deleted key {key.label}")
+
+      if project != None:
+         project.delete()
+         self.logger.info(f"deleted project {project.name}")
+
+   def new_shell(self, ip):
+      shell = spur.SshShell(
+         hostname=ip,
+         username="root",
+         load_system_host_keys=False,
+         missing_host_key=spur.ssh.MissingHostKey.accept,
+         private_key_file=self.private_key_path
+      )
+
+      return shell
+
+   def get_device(self, device_id):
+      return self.packet_manager.get_device(device_id=device_id)
+
+   def get_device_ip(self, device_id):
+      d = self.get_device(device_id)
+      ips = d.ips()
+      self.ip = ips[0].address
+
+      return self.ip
+
+   def delete_device(self, device_id):
```

Comment:
After reading this, `Welder` could easily be used/extended to create a pytest fixture, where things like IP address, `run_ssh_command` could be used in tests.